### DISCUSSION TITLE
Counter variable in foreach loop

### DISCUSSION THREAD

**Discussion by xsoheilalizadeh** [dotnet/csharplang#1584] [Language Ideas]
A feature to define counter variable in `foreach` loop, something like:
```
 foreach (var item in items; int counter = 0) 
 {
     counter++;
  
     Console.WriteLine (counter);
 }
```

currently we can use following way:
```
 var counter = 0;

 foreach (var item in items) 
 {

      counter++;
   
      Console.WriteLine (counter);
 }
```

**Comment by TonyValenti** [comment]
As a more general solution, I'd prefer to see the For-Each solution support blocks in its regions.

````
            for ({ var IE = List.GetEnumerator(); int Index = 0; } ; IE.MoveNext() ; Index++) {
                 Console.WriteLine($"{Index}:  {IE.Current}");
            }
````

**Comment by TonyValenti** [comment]
This is some code that I use too:
````
 public class SelectIndexResult<T> {
        public T Value { get; set; }
        public int Index { get; set; }
    }

    public static class CollectionExtensions {

        public static IEnumerable<SelectIndexResult<T>> WithIndexes<T>(this IEnumerable<T> This) {
            return This.Select((Item, Index) => new SelectIndexResult<T>() {
                Value = Item,
                Index = Index,
            });
        }
}
````

**Comment by ufcpp** [comment]
https://gist.github.com/ufcpp/2b3e1a5821169f6b21ded175ad05c752#file-program-cs-L25

**Comment by svick** [comment]
I think that a tuple-based `Indexed` extension method, like the one shown by @ufcpp, should be added to `Enumerable`. This lets you access the index inside a `foreach`, and does not require any changes to the language.

**Comment by ethanae** [comment]
Why not use `Select` to map a collection which gets the current object and the index passed in? 

`list.Select((element, index) => { ... })`?

https://stackoverflow.com/a/2471611/4716488

**Comment by ufcpp** [comment]
@ethanae 
To avoid allocation, it would be better using struct https://gist.github.com/ufcpp/2b3e1a5821169f6b21ded175ad05c752#file-a-cs-L6

**Comment by jnm2** [comment]
I call mine `AsIndexed()` =)

**Comment by bondsbw** [comment]
@TonyValenti Your `for` syntax might be provided via #377.

**Comment by DavidArno** [comment]
@jnm2,
[Mine is just called `Indexed()`](https://github.com/DavidArno/SuccincT/wiki/IndexedEnumerations) 😄 

As @svick says, it would be good to have the core libraries add this to `IEnumerable<T>` as an extension method, or as a default method implementation when we get that feature.

**Comment by TheUnlocked** [comment]
Counter proposal:
```cs
 foreach (var item in items; in int counter) 
 {
     Console.WriteLine(counter);
     Console.WriteLine(item);
 }
```
This would allow the user to access the counter, but not actually effect it. If they want to manipulate the counter, they should really just use a `for` loop. However, the user _should_ still be able to access the counter.

`readonly` might also be a better keyword instead of `in`. Having no keyword (or an optional keyword) could also work, as long as the counter is still unchangeable.

**Comment by svick** [comment]
@TheUnlocked 

That's confusing, since `foreach` already uses `in` with different meaning.

Also, the iteration variable of a `foreach` loop is `readonly`, but doesn't have any modifier indicating that, so that would be inconsistent.

**Comment by bondsbw** [comment]
Another counter counter-proposal: 😛
```c#
foreach (var item[index] in items) 
{
     Console.WriteLine(index);
}
```

(Edited, we shouldn't be incrementing, that can be handled automatically.)

**Comment by TheUnlocked** [comment]
I agree that `in` was probably not the best choice. However, if it's just `int counter` without any modifier, my concern is that there would be an implication that it's just like any other variable, which wouldn't be true. Requiring `readonly` or `in` would get the message across, though as you pointed out, `in` could be confusing.

The difference between them as I would interpret it on first glance is:
Nothing: Isn't readonly; same variable for the entire loop
`in`: Readonly; same variable for the entire loop
`readonly`: Readonly; different variable for each iteration

The difference between the latter two won't be important in most cases, but could be important in some.

Edit: I suppose as an alternative, it could be up to the user to decide which they want for their use case.

**Comment by TonyValenti** [comment]
I liked the in proposal but i thought it should have been an out var.

On Fri, Jun 1, 2018 at 6:10 PM Unlocked <notifications@github.com> wrote:

> I agree that in was probably not the best choice. However, if it's just int
> counter without any modifier, my concern is that there would be an
> implication that it's just like any other variable, which wouldn't be true.
> Requiring readonly or in would get the message across, though as you
> pointed out, in could be confusing.
>
> The difference between them as I would interpret it on first glance is:
> Nothing: Isn't readonly; same variable for the entire loop
> in: Readonly; same variable for the entire loop
> readonly: Readonly; different variable for each iteration
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/csharplang/issues/1584#issuecomment-394032337>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AM-qVpH-qxxOZwGYc5WO5Yq9OPUHqiMGks5t4cnmgaJpZM4UUvCj>
> .
>
-- 


Always glad to help,

[image: AlphaDrive for Clio] <https://www.alphadriveforclio.com/>

Tony Valenti
Software Engineer
T: 402-401-7568 <javascript:void(0);>
E: Tony.Valenti@AlphaDriveForClio.com
W: www.AlphaDriveForClio.com <https://www.alphadriveforclio.com/>

Schedule a Meeting with Me
<https://www.alphadriveforclio.com/schedule-meeting>
This message contains confidential information and is intended only for the
intended recipients. If you are not an intended recipient you should not
disseminate, distribute or copy this e-mail. Please notify us immediately
by e-mail if you have received this e-mail by mistake and delete this
e-mail from your system. E-mail transmission cannot be guaranteed to be
secure or error-free as information could be intercepted, corrupted, lost,
destroyed, arrive late or incomplete, or contain viruses. Therefore we do
not accept liability for any errors or omissions in the contents of this
message, which arise as a result of e-mail transmission. If verification is
required please request a hard-copy version.

**Comment by jnm2** [comment]
One line of reply, 44 lines of email boilerplate. 😄

**Comment by CyrusNajmabadi** [comment]
Would it not just be sufficient to do this (pseudocode, not tested):

```c#
public static IEnumerable<(T value, int index)> Indexed<T>(this IEnumerable<T> values)
{
    int index = 0;
    foreach(var v in values)
    {
        yield return (v, index);
        index++;
    }
}

// then you can do:

foreach (var (val, idx) in values.Indexed())
{
}
```

?
```

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

> Would it not just be sufficient to do this

(For the sake of argument) that wouldn't work on values that are enumerable via pattern rather than by implementing the `IEnumerable<T>` interface.

With LINQ you can also do:

```cs
foreach (var (item, index) in values.Select((item, index) => (item, index))) {
}
```

**Comment by yaakov-h** [comment]
> that wouldn't work on values that are enumerable via pattern rather than by implementing the IEnumerable<T> interface.

That could work with `SEnumerable` once we get shapes, but would then fall apart if #600 is ever implemented.

**Comment by jnm2** [comment]
@yaakov-h Why would it fall apart?

**Comment by yaakov-h** [comment]
@jnm2 because I forgot about the second half of "Shapes and Extensions"? 😄

**Comment by DavidArno** [comment]
@CyrusNajmabadi,
> Would it not just be sufficient to do this...

Yes, this is sufficient in all but edge cases, [which is why SuccincT does it this way](https://github.com/DavidArno/SuccincT/blob/master/src/SuccincT/Functional/EnumerableExtensions.cs) for example. It would be good to have this added to the BCL, net standard etc though, rather than there being many many near-identical implementations across 3rd party code. But I guess this repo isn't the place to suggest BCL changes.

**Comment by VisualMelon** [comment]
@DavidArno out of interest, what edge cases are you referring to? Perhaps I'm just being slow, but I can't think how those could go wrong.

**Comment by DavidArno** [comment]
@VisualMelon,

As I understand it, the rules around the source object for `foreach` simply requires it support `GetEnumerator()`  and the type returned by that method must have a `Current` property and a `MoveNext()` method. There's no need for it to actually implement `IEnmerable`/`IEnumerable<T>`. So the `Indexed` extension method for `IEnumerable<T>` won't work with all existing `foreach` loops.

It wouldn't break existing code. It just isn't a universal solution.

**Comment by VisualMelon** [comment]
@DavidArno ah, the old `foreach` dodgyness... thanks

**Comment by Thaina** [comment]
@HaloFour I think this

> `values.Select((item, index) => (item, index))`

Is too common and too redundant enough to put into BCL Linq as the @CyrusNajmabadi suggest

Just

```C#
public static IEnumerable<(T value, int index)> Indexed<T>(this IEnumerable<T> values)
{
    return values.Select((item, index) => (item, index));
}
```

Or just

```C#
public static IEnumerable<(T value, int index)> Select<T>(this IEnumerable<T> values)
{
    return values.Select((item, index) => (item, index));
}

foreach(var (item,index) in items.Select()) { }
```

**Comment by AustinBryan** [comment]
@CyrusNajmabadi, @Thaina  I was right about to comment that's the way that I do it. One of the many nices things that can be done with tuples, except I used a `yield return` like @CyrusNajmabadi 

That being said, if they did want to add some syntax that made it even quicker suchas the suggested:

```csharp
foreach (var item[index] in items) { }
```

I wouldn't mind. The bracket notation is cool because it mirrors how we use it, but when I first saw it I got confused what was happening, because I thought we were indexing. 

Maybe one of these:
```csharp
foreach (var item, index in items) { }
foreach (var item in items; index) { }
foreach (var item, var index in items) { }
foreach (var item in items; var index) { }
foreach (var item, int index in items) { }
foreach (var item in items; int index) { }
```
To look more similar to how we can already declare variables via pattern matching. 
The first would be `int` or `long` if needed (but honestly, why is your collection even that big?) 
They wouldn't need to specify the index variable if they won't need it. 

The ones with the `;` look like a more familiar `for` loop.

**Comment by TheUnlocked** [comment]
I'm not sure how I feel about the bracket notation. It sort of makes sense, but that kind of notation for declaring a variable doesn't exist anywhere else in the language. In addition, it hardly fits with iterator functions (there's no indexing there, just `yield return`s).

If the bracket notation is not used, I think there should be a type preceding the variable name, as not having one would differ from every other way to declare a variable in C#.

@AustinBryan I have a question about `var`: Would it just be the same as putting `int`? I would imagine that any integer type (`sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`) would be allowed, and since it's impossible to know how long the iterator would be before it finishes, the C# would need some kind of default if `var` is allowed in this context.

Edit: I suppose I shouldn't necessarily leave out `dynamic` as a valid type for the index variable.

**Comment by AustinBryan** [comment]
@TheUnlocked I guess you're right. I mean, in nearly every single case it's going to end up being an `int`. I highly doubt anyone would need `long` and I heard that `int` is faster than `short` anyways, because everyone used it so it's been heavily optimized.

**Comment by jnm2** [comment]
If speed matters, `long` is faster than `int` on x64. I'm thinking once the proposal goes through for C# to expose the CLR's native int number type, folks will start using that for loops and things.

**Comment by DavidArno** [comment]
@jnm2, personally I use `int` as it’s the easiest one to type. `nuint` seems a bit of a clunky thing to use instead. I guess it depends on whether one micro-optimises or not. 😜

**Comment by jnm2** [comment]
@DavidArno it would be `nint` vs `var` for me.

**Comment by sirgru** [comment]
Other languages have this useful feature:
Go:
``` Go
for index, element := range someSlice {
    // index is the index where we are
    // element is the element from someSlice for where we are
}
```
Swift:
```Swift
for (index, element) in list.enumerated() {
  print("Item \(index): \(element)")
}
```
D:
```C
foreach (i, e; [4, 5, 6]) {
    writeln(i, ":", e);
}
// 0:4 1:5 2:6
```
They all put the indexer before the variable which I find weird. I like this syntax best:
```C#
	foreach(var item @ index in collection ) {
		...
	}
```
The extension methods on IE have negative performance characteristics just for one indexer. I think this should be a standard language feature.

**Comment by CyrusNajmabadi** [comment]
> The extension methods on IE have negative performance characteristics just for one indexer

Why is that the case?   

If you have:

```c#
public struct IndexedEnumerable<T> 
{
    private readonly IEnumerable<T> values;

    public IndexedEnumerable(IEnumerable<T> values)
        => this.values = values;

    public Enumerator GetEnumerator()
        => new Enumerator(values.GetEnumerator());

    public struct Enumerator
    {
        private readonly IEnumerator<T> enumerator;
        public int index;

        public Enumerator(IEnumerator<T> enumerator)
        {
            this.enumerator = enumerator;
            index = -1;
        }

        public bool MoveNext()
        {
            index++;
            return enumerator.MoveNext();
        }

        public (int, T) Current => (index, enumerator.Current);
    }
}

public static class Enumerable
{
    public static IndexedEnumerable<T> Indexed<T>(this IEnumerable<T> values)
        => new IndexedEnumerable<T>(values);
}


class Test
{
    static void Main(string[] args)
    {
        foreach (var (index, val) in args.Indexed())
        {

        }
    }
}
```

Then that should be pretty darn fast.  There are no actual allocations going on here.  

--

For those concerned about what happens if you have an existing struct-enumerator to work with, then Shapes helps solve that becuase you'd write the above with SEnumerable and SEnumerator instead, giving you a shape-based way to do this with almost no perf overhead.

So, in other words, i don't think we actually need this proposal, because it can be done today already with existing langauge features (or subsumed into other upcoming language features).

**Comment by CyrusNajmabadi** [comment]
> They all put the indexer before the variable which I find weird. I like this syntax best:

That makes sense to me.  If i iterate a dictionary, i get Key-Values, not Value-Keys.  For a sequence, the key is the index.  So i get Index-Value, not Value-Index.

**Comment by CyrusNajmabadi** [comment]
Note: if we ever get struct-enumeration in the language, tehn the above could just be written as:

```c#
public static class Enumerable
{
  public static struct IndexedEnumerable<T> Indexed<T>(this IEnumerable<T> values)
  {
    int index = 0;
    foreach (var val in values)
    {
      yield return (index++, val);
    }
}
```

But until then, manually writing it out is the way to go.

**Comment by sirgru** [comment]
> That makes sense to me. If i iterate a dictionary, i get Key-Values, not Value-Keys. For a sequence, the key is the index. So i get Index-Value, not Value-Index.

I get the logic. I guess it makes more sense when `value` is a tuple, so it's like a KVP. However, my logic was: there are plenty of cases where index doesn't logically correspond to a Key, for example when iterating dictionaries themselves (which wouldn't make much sense but still). If the indexer comes later like in my syntax it feels in some way like an optional parameter - the optional stuff comes last. So there is a foreach like it is today and the foreach with an indexer which is optional. It also reads much better imo.

> Then that should be pretty darn fast. There are no actual allocations going on here.

I agree. But my understanding is with simple collections such as lists and arrays the compiler makes optimizations to not threat them through an IE interface, instead iterates over them like in a for loop, where foreach becomes only syntactic sugar. In this case such optimization cannot be made.

**Comment by CyrusNajmabadi** [comment]
> I agree. But my understanding is with simple collections such as lists and arrays the compiler makes optimizations to not threat them through an IE interface, instead iterates over them like in a for loop, where foreach becomes only syntactic sugar. In this case such optimization cannot be made.

I mentioned this bit:

> For those concerned about what happens if you have an existing struct-enumerator to work with, then Shapes helps solve that becuase you'd write the above with SEnumerable and SEnumerator instead, giving you a shape-based way to do this with almost no perf overhead.

> So, in other words, i don't think we actually need this proposal, because it can be done today already with existing langauge features (or subsumed into other upcoming language features).

**Comment by ekolis** [comment]
Throwing my hat into the ring for a syntax suggestion:

```
var list = new string[]{"a", "b", "c"};
foreach (var q, i in list)
{
    Console.WriteLine($"item {i} is {q}");
}
```

**Comment by theunrepentantgeek** [comment]
Issue #3194 would enhance they way **foreach** looks for an implementation pattern, in the same way that **asyc** and deconstruction is already pattern based.

If that lands in C#, a simple extension method would make iteration with an index trivial, no other new syntax for C# required.

First, write a suitable extension method:

``` csharp
public static IEnumerator<(T, int)> GetEnumerator<T>(this IEnumerable<T> items)
{
    int index = 0;
    foreach(var i in items)
    {
        yield return (i, index++);
    }
}
```

and then you can iterate over any sequence using a tuple:

``` csharp
foreach ( var (Person, int) in FindPeople()) { ... }
```
or, perhaps preferably,
``` csharp
foreach ((Person p, int i) in FindPeople()) { ... }
```

An independent enhancement to the compiler to make the implementation of `GetEnumerator()` based on a struct may event allow this to be allocation free.

**Comment by 333fred** [comment]
Fwiw @YairHalberstadt has an implementation pr out that is in the process of being reviewed, and if everything goes nicely I hope it'll be in 9. No guarantees, but that's my hope.

**Comment by YairHalberstadt** [comment]
@theunrepentantgeek I don't think that will work, since to preserve backwards compatibility IEnumerable is preferred over an extension GetEnumerator, and besides it is not target typed.

**Comment by theunrepentantgeek** [comment]
That's a shame. 

When I skimmed through #3194, I thought it would fall back to the extension `GetEnumerator()` if using the one built into the type didn't compile, allowing this scenario.

Apologies for misreading it.

**Comment by Ramobo** [comment]
How about this slightly modified version of the original proposal?
```csharp
foreach (Foo foo in foos; index) // `index` is implicitly a read-only `int`. Allow custom name for the case of nested loops, and for consistency with `for`.
{
    Console.WriteLine($"[{index}] {foo}");
}

// Translates to:
using (IEnumerator<Foo> enumerator = foos.GetEnumerator())
{
    int index = 0;

    while (enumerator.MoveNext())
    {
        Foo foo = enumerator.Current;

        Console.WriteLine($"[{index}] {foo}");

        index++;
    }
}
```

**Comment by seanblue** [comment]
What would it take to add the tuple foreach syntax into the language itself instead of having to rely on an extension method? 

**Extension method example from above:**
```c#
foreach (var (item, index) in collection.Indexed())
{
    DoSomething(item, index);
}
```

**Proposal:**
```c#
foreach (var (item, index) in collection)
{
    DoSomething(item, index);
}
```

I'd like the language itself to contextually return `T` or a tuple `(T, int)` in the foreach simply based on which `var` (or hard type) is declared to the left of the `in`.

I think this approach is most consistent with current C# functionality because it's the same syntax used in `Select` and extension methods like above.

**Reply by Thaina** [reply]
I would say anything it take does not worth it. This feature could be solved easily with a little extension method. So changing the compiler just for that is absolutely overkill

Not to mention it was not intuitive that `foreach` can provide index if you just put a tuple at it. And it ambiguous that the object was the tuple itself or you want index of it

**Reply by HaloFour** [reply]
It would take the language design team to agree, and so far they don't.

Note that your proposal would be ambiguous if the enumerator already returned a tuple.  It also wouldn't work for non-generic enumerables.

**Reply by seanblue** [reply]
> Note that your proposal would be ambiguous if the enumerator already returned a tuple.

@HaloFour I don't think that's the case. The compiler would have to detect the actual generic type. If the `IEnumerable`'s generic type is already (for example) `(string, int)`, then the foreach could loop over `(string a, int b)` or `((string a, int b), int index)`. I think that's pretty understandable, personally.

> It also wouldn't work for non-generic enumerables.

This part I can't really comment on though. I'm not sure what all the limitations are for non-generic enumerables because I almost always avoid using them. If this makes the proposal unworkable, so be it.

> This feature could be solved easily with a little extension method. So changing the compiler just for that is absolutely overkill

@Thaina To be fair though, a significant portion of new C# features are syntactic sugar that don't actually add new functionality, but do (depending on your view) make things more convenient.

**Reply by Thaina** [reply]
@seanblue
> new C# features are syntactic sugar that don't actually add new functionality

Every feature is syntactic sugar. But most of them provide high level of convenient without great alternative

My point is 

> This feature could be solved easily with a little extension method

This alternative is not complex enough to require introducing a new language feature just to solve it

If, as in the olden day we don't have tuple and tuple deconstruction nor extensoin method, I might consider that any alternative is not appeased. But now we already have all of those and just 8-10 characters could be saved is not important enough to introduce compiler change in the language itself

**Comment by Thaina** [comment]
Another benefit of using extension method over syntax is you can chain it with linq and clearly understand what it is

```C#
// suppose there is Select that return (item,index)
foreach(var ((item,i),count) in list.Select().Where((pair) => pair.item.IsWhatIWant).Select())
{
    // i is original index in the list
    // count is number of filtered/sorted
}
```

**Comment by cheong00** [comment]
I think we should do the change on for loop rather then foreach loop because not all IEnumerator interface implementations implements .Item[Index] as member (say, [Hashtable](https://docs.microsoft.com/en-us/dotnet/api/system.collections.hashtable.item?view=net-5.0) or most key-based collections) 

And for collections like Tree which we can have multiple transversal methods (and typically implements IEnumerator), index doesn't really have meaning there.
